pkg/reconciler/managed/metrics.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package managed
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	kmetrics "k8s.io/component-base/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+func init() { //nolint:gochecknoinits // metrics should be registered once
+	metrics.Registry.MustRegister(drift, mr, mrReady, mrSynced, mrDetected, mrReadyDuration, mrDeletionDuration)
+}
```

Comment:
I've already tried this pattern. As you suggested the best (maybe the only) place to initialize and register the metrics is the main function which comes from Upjet for official providers. Passing through these metrics to the reconciler is not straightforward and will make things more complicated than it should be. We also need to think about other providers which are not generated by Upjet. If we decide to follow this pattern, I guess following steps need to be done:
- Define which metrics to have in crossplane-runtime
- Merge and consume that change in Upjet
- Initialize and register those metrics in Upjet and pass them to the controllers (I'm still not sure how complicated this will be)
- In MR reconciler, make sure that those metrics are passed and record metrics.

I'm not sure what we would achieve with this is worth it.

Reply:
I see it as adopting the same pattern we already have for other instrumentation - e.g. loggers and event handlers. We initialize these in main and plumb them down to the reconciler implementation.

Yes, this is more effort, but I feel it's worth it. A _library_ like crossplane-runtime should offer you instrumentation options but shouldn't force choices on you, for example by forcing a specific logger or metrics library inherently because you imported a package.

Passing instrumentation implementations as dependencies also allows us to pass in stub/fake implementations during testing if we want to, for example if we wanted to test that we were emitting logs or metrics correctly.

Some interesting reading on the topic:

* https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html
* https://dave.cheney.net/2017/06/11/go-without-package-scoped-variables
* https://github.com/prometheus/docs/issues/506
* https://groups.google.com/g/prometheus-developers/c/l5dXdTkJtl0

It's worth noting that Prom maintainers (e.g. Brian and Björn) are split on this. I'm more aligned with Björn. Perhaps if this wasn't a library the path of least resistance - using global state - would be fine. In this case, it is not.

Reply:
I've switched to the pattern you suggested.